ucsb-cs156/proj-courses
# Staff f22 pc jobs frontend

# Overview

In this PR we add the frontend code for the background jobs system (i.e. asynchronous long running jobs).

Async Jobs allow our application to do things that take longer than a single web request/response transaction (i.e. things that take longer than a second or two).  Most web servers will timeout if a request/response takes longer than some threshold, ranging somewhere between 1 second and 5 seconds.   The timeouts typically don't happen when testing on localhost, but do happen when deployed on real web servers.  

Another PR introduced the async job backend.  This PR fixes a couple of small backend problems, as well as adding the frontend components.

# User Story

As an admin and/or developer
* I can launch async jobs
* So that we can add more new features to the app that take longer than a second to execute, such as loading data

# How to Test

1. Login as an admin
2. Find the Manage Jobs item on the Admin menu
3. Click to open the Test Job panel
4. Submit a job with fail=false, and sleepMs=4000
5. You should see that job appear in the job list
6. After 4 seconds it's status should change to complete and log output should be updated.

Try this again with fail=true, and various values of sleepMs including 0

# A Guide to the Code

There's a lot going on in this PR; sorry for the size of it.  Here's an overview of the changes to the code:

Frontend:
* App.js: We added a route for the frontend Admin Jobs page
* jobsFixtures.js: fixtures for the backend route that handles jobs
* JobComingSoon.js: a component that's a placeholder for future jobs in the frontend
* JobsTable.js: a component for displaying information from jobs
* TestJobForm.js: a form for filling in the information that we have to submit to the test job (fail true/false, and the number of milliseconds to sleep)
* AppNavbar.js: we added a link to the AdminJobsPage
* OurTable.js: we added two new column types: 
  - a date column type that automatically handles time zone information and formatting and renders a date/time in a readable format in the local time zone
  - a column type for preformatted raw ASCII text (for the logs), uses `Plaintext.js` below
* Plaintext.js: a component for raw preformatted ASCII text (for the logs)
* useBackend.js: some adjustments to error handling to make it more robust and testable
* stories: stories for new components listed above
* tests: tests for changes listed above

Backend:
* CoursesApplication.java: Added an annotation on the main class, `@EnableJpaAuditing(dateTimeProviderRef = "utcDateTimeProvider")` to enable th createdAt and updatedAt fields on entities to be automatically populated, and a method to provide the date with a timezone
* Job.java: changed LocalDateTime to ZonedDateTime so that job dates display in the correct time zone (they are supposed to be stored in UTC but displayed in localtime)
* Many controller tests and service tests; just needed `@AutoConfigureDataJpa` added so that the bean to provide date/time for JPA operations is provided in the test environment
